core/emulate.c

```diff
@@ -823,9 +850,19 @@ static void em_mov(struct em_context_t *ctxt)
 
 static void em_movzx(struct em_context_t *ctxt)
 {
-    uint64_t value = 0;
-    memcpy(&value, &ctxt->src1, ctxt->operand_size);
-    ctxt->dst.value = value;
+    ctxt->dst.value = 0;
+    memcpy(&ctxt->dst.value, &ctxt->src1.value, ctxt->src1.size);
+}
+
+static void em_movsx(struct em_context_t *ctxt)
+{
+    uint64_t extension = 0;
+    if (ctxt->src1.value & (1 << ((8 * ctxt->src1.size) - 1))) {
```

Comment:
The `<<` operation may cause an int overflow when `ctxt->src1.size == 8`. Two possible ways to fix that:

1. `1 <<` => `1ULL <<`
1. `if (ctxt->src1.value >> (8 * ctxt->src1.size - 1))`

Reply:
Sorry I overlooked this edge case, `movsx` was defined so that `ctxt->src1.size` can only be *1* or *2*.
This is fixed now. I have picked your first suggestion (`1 <<` => `1ULL <<`), as it will work regardless of whether the higher bits (above the sign bit) are zero.

Reply:
Ah, right, it's the source operand. Sorry about the fuss!